pkg/controller/cloud/node_controller.go

```diff
@@ -271,6 +297,12 @@ func (cnc *CloudNodeController) MonitorNode() {
 					}
 				}(node.Name)
 
+			} else {
+				// if taint exist remove taint
+				err = controller.RemoveTaintOffNode(cnc.kubeClient, node.Name, node, ShutDownTaint)
```

Comment:
This means that the semantics or this Taint are somewhat cloudy. 
- It's applied when Node is NotReady and VM is in a given state
- It's removed when Node is Ready

It may be correct, but I guess full negation would be better, i.e.
- It's removed when Node is Ready or VM is not in the given state

Reply:
You are correct that it would be better to update that all the time. However, if we do that we will make request to cloudprovider api every 5 seconds. 99% of these requests are not needed. State is not changed. If we have 100 node clustee it maked 100 api requests every 5 second. Imo too mich because its not really needed

Reply:
Instead of asking about every single VM we can just ask for the list of all VMs and work with that.

Reply:
@zetaab what do you mean we will make it every 5 seconds? Nothing I see Is on such a short interval.

Reply:
@jhorwit2 monitornode function is executed every 5 seconds

Reply:
That seems unnecessarily quick. The KCM marks nodes not ready after 40 seconds by default, so i'm not sure why this would need to be faster than that.

Reply:
@zetaab To your point earlier, this would cause up to 200 api calls every 5 seconds to the cloud provider since you'd not only check for if it's shutdown but also check if it's been deleted for every node that's healthy.

Reply:
I think we could do this whole thing little bit better. If we increase monitornode time, it will mean that removing / adding shutdown taint takes 40seconds+monitornode delay maximum. I mean that if node notready takes 40seconds, it might mean that adding taint takes 40seconds+5seconds maximum currently.

What about if we re-structure this whole feature, if node goes notready state why it cannot push message towards monitornode(using channel maybe)? Then there is (almost) no delay, and we do not need refresh that 24/7 every 5seconds?

Reply:
@jhorwit2 yep 200 maximum per 5seconds with 100 nodes. @gmarek well we could do that. However, I would like to modify then this whole monitornode function. Instead of checking nodes status one by one in cloudprovider, we should make bulk query towards cloudprovider (which input contains all nodes providerids). Then cloudprovider queries providerids statuses from cloudprovider API, and returns three arrays as return to this monitornode function. These arrays are: shutdown, active and removed. Where shutdown means = add taint if taint not exist yet. Active = if taint exist remove taint, removed = delete node